wiring/src/spark_wiring_ble.cpp

```diff
@@ -577,81 +576,76 @@ BleAdvertisingData::BleAdvertisingData(const iBeacon& beacon)
 
 size_t BleAdvertisingData::set(const uint8_t* buf, size_t len) {
     if (buf == nullptr || len == 0) {
-        selfLen_ = 0;
-        return selfLen_;
+        selfData_.clear();
+        return selfData_.size();
     }
-    len = std::min(len, (size_t)BLE_MAX_ADV_DATA_LEN);
-    memcpy(selfData_, buf, len);
-    selfLen_ = len;
-    return selfLen_;
+    selfData_.clear();
+    len = std::min(len, (size_t)BLE_MAX_SCAN_REPORT_BUF_LEN);
+    selfData_.append(buf, len);
+    return selfData_.size();
 }
 
 size_t BleAdvertisingData::set(const iBeacon& beacon) {
     clear();
 
     if (beacon.UUID().type() == BleUuidType::SHORT) {
-        return selfLen_;
+        return selfData_.size();
     }
 
-    selfData_[selfLen_++] = 0x02;
-    selfData_[selfLen_++] = BLE_SIG_AD_TYPE_FLAGS;
-    selfData_[selfLen_++] = BLE_SIG_ADV_FLAGS_LE_ONLY_GENERAL_DISC_MODE;
-    selfData_[selfLen_++] = 0x1a; // length
-    selfData_[selfLen_++] = BLE_SIG_AD_TYPE_MANUFACTURER_SPECIFIC_DATA;
+    selfData_.append(0x02);
```

Comment:
Any append in theory can fail. Not sure how exactly we want to deal with allocation failures here. Reserving `BLE_MAX_ADV_DATA_LEN` might be a good idea just in case as well. cc @XuGuohui

Reply:
Added a `reserve` for `BLE_MAX_ADV_DATA_LEN` before all the appends, returning 0 if it fails (since the `set` function doesn't return negative). Also in `advertise(iBeacon)` added a check that the length of the advert is larger than 0.

Reply:
You just added a reserve here only. It might fail to append wherever `append()` is used. Adding a reserve is one approach, but it still waste ram. Here are two candidates I can put up:

1. Reserve `BLE_MAX_SCAN_REPORT_BUF_LEN` ram before appending and simply return if it fails. If succeeds, use `trimToSize()` to free unused ram.
2. New a temp vector and fill it with the new data. If it fails at certain append, we abort the data, otherwise, we finally append the temp vector to the `selfData_` vector.

Reply:
Added `trimToSize()` after done appending.